exam3.cpp

```diff
@@ -0,0 +1,478 @@
+        this->codigo =_codigo;
+        this ->nombre = _nombre;
+        this -> precio = _precio;
+        this ->stock = _stock;
+    }
+
+    virtual double calcularPrecioFinal()=0;
+
+    void reducirStock(int cantidad)
+    {
+        if(cantidad >stock)
+        {
+            throw StockInsuficienteException();
+        }
+        stock-=cantidad;
+    }
+
+    virtual void mostrar()
+    {
+        cout << " codigo "<< codigo << endl;
+        cout << "El precio final sera "<< calcularPrecioFinal() << endl;
+        cout << "El stock restante es " << stock <<endl;
+    }
+
+    string getCodigo()
+    {
+        return codigo;
+    }
+    string getNombre()
+    {
+        return nombre;
+    }
+    double getPrecio()
+    {return precio;}
+
+    double getStock()
+    {
+        return stock;
+    }
+};
+
+// Valorador para usar con ArrayGenerico<Producto>::obtenerMayor / obetenerMenor
+double valorProducto(Producto* p)
+{
+    return p->calcularPrecioFinal();
+}
+
+class ProductoNormal : public Producto
+{
+    private:
+    double total;
+    public:
+
+    ProductoNormal(string _codigo, string _nombre , double _precio, int _stock) : Producto(_codigo,_nombre,_precio,_stock)
+    {
+        this -> total = 0;
+    }
+
+    double calcularPrecioFinal() override
+    {
+        total = precio;
+        return total;
+    }
+    void mostrar() override
+    {
+        Producto::mostrar();
+    }
+};
+
+class ProductoConDescuento : public Producto
+{
+    private:
+    double total;
+    double porcentajeDescuento;
+    public:
+
+    ProductoConDescuento(string _codigo, string _nombre , double _precio, int _stock,double _porcentajeDescuento) : Producto(_codigo,_nombre,_precio,_stock)
+    {
+        this -> total = 0;
+        this -> porcentajeDescuento = _porcentajeDescuento;
+    }
+
+    double calcularPrecioFinal() override
+    {
+        total = precio*(1-porcentajeDescuento/100);
+        return total;
+    }
+
+    void mostrar() override
+    {
+        Producto::mostrar();
+    }
+};
+
+
+class ProductoImportado : public Producto
+{
+    private:
+    double total;
+    double impuestoImportacion;
+    public:
+
+    ProductoImportado(string _codigo, string _nombre , double _precio, int _stock,double _impuestoImportacion) : Producto(_codigo,_nombre,_precio,_stock)
+    {
+        this -> total = 0;
+        this -> impuestoImportacion = _impuestoImportacion;
+    }
+
+    double calcularPrecioFinal() override
+    {
+        total = precio*(1+impuestoImportacion / 100);
+        return total;
+    }
+    void mostrar() override
+    {
+        Producto::mostrar();
+    }
+};
+
+class DetalleVenta 
+{
+    private:
+    int cantidad;
+    double subTotal;
+    Producto* producto;
+
+    public:
+
+    DetalleVenta(Producto* _producto, int _cantidad)
+    {
+        this->producto = _producto;
+        this->cantidad=_cantidad;
+        this->subTotal=_producto->calcularPrecioFinal() *_cantidad;
+    }
+
+    string getCodigo()
```

Comment:
Nombres descriptivos

```suggestion
    string getCodigoProducto()
```